solutions/platinum/cses-2429.mdx

```diff
@@ -0,0 +1,115 @@
+---
+id: cses-2429
+source: CSES
+title: Grid Completion
+author:
+---
+
+# Solution
+
+Time complexity: $\mathcal O(N^3)$
+
+<LanguageSection>
+
+<CPPSection>
+
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+typedef long long ll;
+const ll MOD = 1e9 + 7;
+const int maxN = 505;  // Maximum grid size
+
+int N, p[maxN], q[maxN], C[6];
+bool inp[maxN], inq[maxN];
+char S[maxN];
+ll fact[maxN], inv[maxN];
+
+// Compute modular inverse using Fermat's Little Theorem
+ll inverse(ll x) {
+	ll res = 1;
+	ll b = MOD - 2;  // Theorem: x^(MOD-2) ≡ x^(-1) (mod MOD)
+	while (b) {
+		if (b & 1) res = (res * x) % MOD;
+		x = (x * x) % MOD;
+		b >>= 1;
+	}
+	return res;
+}
+
+// Function to compute binomial coefficient C(x, y) modulo MOD
+ll choose(int x, int y) { return (fact[x] * inv[y] % MOD) * inv[x - y] % MOD; }
+
+// Precompute factorials and their modular inverses
+void init() {
+	fact[0] = inv[0] = 1;
+	for (int i = 1; i <= N; i++) {
+		fact[i] = (fact[i - 1] * i) % MOD;
+		inv[i] = (inv[i - 1] * inverse(i)) % MOD;
```

Comment:
doesn't matter all that much but you can calculate inverse factorials without constantly calculating mod inverse

idea is to calculate mod inverse of the biggest factorial, then iterate in reverse and multiply stuff out. lmk if u need more details

Reply:
^ see https://usaco.guide/gold/combo?lang=cpp#method-2-factorial-definition-modular-inverses---mathcalon--log-mod